CMS: Unify system names

jlledom opened the pull request:
**What this PR does / why we need it**:

This is a simpler approach to https://github.com/3scale/porta/pull/3163.

In that PR I added validations to all CMS models so `system_name`:

1. Is always present.
2. Meets the format `%r{\A\w[\w\-/_]+\z}`.

The problem is that there are some sections and pages in the DB that doesn't meet the format, so adding validations would require a backfill migration to fix those two tables.

In this PR, I applied the minimal changes possible to ensure those two conditions above are meet **for all new records**, but it doesn't add the validations, so the existing records not meeting the conditions are allowed in order to avoid breaking anything.

**Which issue(s) this PR fixes** 

Part of [THREESCALE-8991](https://issues.redhat.com/browse/THREESCALE-8991)

**Verification steps** 

- Creating a new section which title contains spaces should generate a system name without spaces.
- Creating a Pages without a system name should generate one automatically, with the proper format.
- Creating a new tenant should generate valid system names for all the default Developer Portal resources.

**Special notes for your reviewer**:

There are changes only in `CMS::Page` and `CMS::Section`, all other models were already fixed, this tables summarizes the state of system names before this PR.

|Type|Required|Format enforced|Value|
|--|--|--|--|
|Section|Yes|No|Generated from `title` parameter|
|Page|No|Yes|From `system_name` parameter|
|Builtin Page|Yes|Yes|From `system_name` parameter|
|Layout|Yes|Yes|From `system_name` parameter|
|Partial|Yes|Yes|From `system_name` parameter|
|Builtin Partial|Yes|Yes|From `system_name` parameter|
|File|No|No|There's no  `system_name`  column in the table|

As the table shows, `system_name` is not applicable for **Files**; while **Builtin Page**, **Page**, **Layout**, **Partial** and **Builtin Partial** already meet the two conditions. So only changes in **Page** and **Section** are required.

josemigallas reviewed (changes requested):
It's missing some tests

akostadinov commented:
Looks good to me. Probably it already has some tests for `create`, right? But it wouldn't for `update` though. Is `system_name` mutable? Do we need the `on :update`?

jlledom commented:
> Looks good to me. Probably it already has some tests for `create`, right? But it wouldn't for `update` though. Is `system_name` mutable? Do we need the `on :update`?

There were no tests for `system_name` autofill AFAIK. I've added some. `system_name` is mutable, but it will be autofilled only when it's empty.

- For sections: it should never be empty because it's presence is validated. So `#update` with an empty `system_name` should never happen, but it's a little effort to make the autofill work also for `#update` so I added it, just in case is there any invalid section in the DB or the customer sends and empty `system_name` through the API when updating.
- For pages: `system_name` is not required, so it's good to add the autofill also for `#update`. When some old page without `system_name` is updated, it will get one.

josemigallas commented on test/unit/cms/page_test.rb:
```diff
@@ -97,4 +97,48 @@ def setup
+    assert_equal 'new-page', page.reload.system_name
+  end
+
+  test "doesn't generate a system_name when one is given" do
```
What about if given an invalid system name, such as `'sysname 1'`?

jlledom replied:
It will return a validation error because the `system_name` format for pages is enforced in the parent class `CMS::Template` https://github.com/3scale/porta/blob/1ece547ea9568f4ed9c845ea477dfaa7b1e451c7/app/models/cms/template.rb#L19-L20

josemigallas replied:
Yes, I mean what about writing a test for that scenario.

jlledom replied:
Done

jlledom commented on test/unit/cms/page_test.rb:
```diff
@@ -97,4 +97,48 @@ def setup
+    page.system_name = ''
+    page.save!
+
+    assert_equal 'new-page', page.reload.system_name
```
No, since the system name is empty, it generates a new one based on the title

josemigallas replied:
But it's updating an existing record, already with a system_name. Maybe I got it wrong, but updating with an empty string means either erase the field or leave it alone.

jlledom replied:
It sets `system_name` to an empty string, yes, that's why it works.
